python/cudf_polars/cudf_polars/dsl/ir.py

```diff
@@ -212,6 +226,17 @@ def evaluate(self, *, cache: MutableMapping[int, DataFrame]) -> DataFrame:
             *(child.evaluate(cache=cache) for child in self.children),
         )
 
+    @property
+    def parts(self) -> PartitionInfo:
+        """Return Partitioning information for this IR node."""
+        try:
+            return self._parts_info
+        except AttributeError:
+            from cudf_polars.experimental.parallel import ir_parts_info
+
+            self._parts_info = ir_parts_info(self)
+            return self._parts_info
```

Comment:
*nit*: I would _prefer_, I think, a pass that adds partitioning information.

We also need to include `parts_info` in the hash, I suspect.

So, sketch:

```
# on IR
_partitioning : PartitionInfo
def get_hashable(self):
    hashable = super().get_hashable()
    partitioning = getattr(self, "_partitioning", None)
    return (hashable, partitioning)

def is_equal(self, other):
    ... # same deal, need to check that partitioning is also equal
```

And then,

```
def add_partitioning(ir: IR) -> IR:
    children = [add_partitioning(c) for c in ir.children]
    copy = ir.reconstruct(children)
    copy._partitioning = ...
    return copy
```
Or similar, I haven't looked exactly how partitions are added.

Reply:
Okay, yeah. I *think* we can do an explicit pass over the IR graph to add partitioning information, but I'm not sure if it's worth the complexity.

In my mind (which may be confused), the mapping between an IR node and its partitioning is deterministic, and fully dependent on the class of the node and its children. If we get the design "right", we shouldn't need to include any partitioning information in the hash (because the non-child arguments and children already should already produce deterministic partitioning).

**Disclaimer:** I'm still very open to other solutions, but this is the direction I'm currently leaning...

Task-graph procedure I currently have in mind:

1. We start with the original (single-partition) translated `ir: IR` graph.
  - If `ir_parts_info` is called on any node in this graph, it would return a simple single-partition `PartitionInfo` object.
2. We call `lower_ir_graph` to replace the IR nodes with specialized sub-classes for multi-partition execution.
  - These subclasses would usually be thin aliases. For example `PartwiseSelect` may correspond to a `Select` node that simply preserves the partitioning of its children.
  - We will replace `Scan` with something like `ParallelScan` or `ReadParquet`. The `ir_parts_info` dispatch for `Scan` will return a simple single-partition `PartitionInfo` object.  However, `ir_parts_info(<ParallelScan>)` would return multiple partitions (for a large-enough dataset).
  - We would also replace `GroupBy` with something like `GroupByFinalize(GroupByReduce(GroupByParts(...)))` (if the child node has multiple partitions).
3. (Optional) We add additional rewrite passes to improve multi-gpu performance (if/when necessary).
  - This may include fusion, for example.
4. We traverse the new IR graph, and generate the final task graph.
5. Schedule/execute the task graph.

Given this design, a unique IR object should always return unique partitioning information, and so it doesn't make much sense to "attach" partitioning information.

Reply:
This logic was moved (https://github.com/rapidsai/cudf/pull/17262#discussion_r1842672978)